### ISSUE TITLE
Feature Request - Allow listen on ipv6

### ISSUE THREAD

**Issue by pixelmaven** [typesense/typesense#579]
## Description

It looks like it's currently not possible to have Typesense listen on an ipv6 address? I tried a few different addresses without luck, also applied on via both env vars `TYPESENSE_API_ADDRESS` and `TYPESENSE_LISTEN_ADDRESS`.

## Steps to reproduce

```
TYPESENSE_API_ADDRESS = "::"
```

## Expected Behavior

Expect Typesense http server to listen on an ipv6 address.

## Actual Behavior

```
Failed to connect to *** port 8080: Connection refused
```

## Metadata

**Typesense Version**: 0.22.2

**OS**: Docker on Ubuntu

**Comment by maxostarr** [comment]
Is there potentially a quick patch for this issue? I have an ipv6 only k8 cluster that I'm trying to run Typsense in.

**Comment by jcrapuchettes** [comment]
Is this still an issue?

**Comment by kishorenc** [comment]
Yes, we don't support listening on ipv6 yet. The listener currently only explicitly uses `AF_INET `. https://github.com/typesense/typesense/blob/main/src/http_server.cpp#L148

Happy to accept a patch.

**Comment by jcrapuchettes** [comment]
Sadly, that isn't the only location that will need to be updated. If I'm correct, the butil package used for `ip2str` and `hostname2ip` are only `AF_INET`.

**Comment by maxostarr** [comment]
Got it! I've managed to get around this by using an NGINX sidecar container to 'translate' the IPV6 traffic to IPV4 for the typesense server. Thanks for the insights!

**Comment by jcrapuchettes** [comment]
Update: I was able to get Typesense to work in an IPv6-only cluster by using a number of `6tunnel` tunnels from localhost. `6tunnel` can be installed via apt in debian and ubuntu. The trick for clustering was to configure Typesense like it was running 3 nodes on the same instance (each using a different set of ports) and then use `6tunnel` to forward the traffic to the correct node. In total, each node has 7 tunnels.
```bash
# Listen on IPv6 ports 8108 and make available via the defined api port for the node. For External Traffic
6tunnel -6 -l :: 8108 localhost $api_port
# Listen on IPv6 ports $api_port and make available via the defined api port for the node. For Cross-Node Traffic
6tunnel -6 -l <private v6 ip> $api_port localhost $api_port
6tunnel -6 -l <private v6 ip> $peering_port localhost $peering_port

# API and Peering tunnels for other nodes
6tunnel -4 -l localhost <node 2 api port> <node 2 private v6 ip>
6tunnel -4 -l localhost <node 2 peering port> <node 2 private v6 ip>
6tunnel -4 -l localhost <node 3 api port> <node 3 private v6 ip>
6tunnel -4 -l localhost <node 3 peering port> <node 3 private v6 ip>
```

**Comment by uncvrd** [comment]
@jcrapuchettes When you have a second, could you also share your `typesense` command to start each node? I'm having some issues with my peering address and not sure if I should be using the IPv4 or IPv6 address (im trying to use Fly.io)

**Comment by jcrapuchettes** [comment]
> @jcrapuchettes When you have a second, could you also share your `typesense` command to start each node (including the nodes file)? I'm having some issues with my peering address and not sure if I should be using the IPv4 or IPv6 address (im trying to use Fly.io)

`/opt/typesense-server --api-port $api_port --peering-address 127.0.0.1 --peering-port $peering_port`

**Comment by samhenderson511** [comment]
I'm struggling with this atm as I'm trying to deploy a cluster on Fly.io, is this likely to be added in v29?

**Comment by jcrapuchettes** [comment]
> I'm struggling with this atm as I'm trying to deploy a cluster on Fly.io, is this likely to be added in v29?

@samhenderson511 I just did preliminary testing on my computer the `typesense/typesense:30.0.rc1` image and it has the IPv6 support. We are going to be testing it in the near future.

**Comment by kishorenc** [comment]
Thanks for helping ship this and sorry it took so long to merge.

**Comment by alangmartini** [comment]
Available in v30 forward
